### ISSUE TITLE
Roadmap of YOLO-World

### ISSUE THREAD

**Issue by wondervictor** [AILab-CVC/YOLO-World#109]
**This issue will be kept open and pinned for a long time, as we hope to hear everyone's opinions, suggestions, and needs!**
We want to make YOLO-World stronger and encourage more diverse applications, especially practical ones. We maintain an open and free attitude. YOLO-World is currently in active development and improvement, and we are trying our best to do well in upstream pre-training and downstream deployment tools. At present, our manpower is limited, so we hope you can give us some time and contribute your experience or help when you can!

If you have a good idea or need, just reply to this issue and @ me. I will respond promptly when I see it, and consider adding it to the TODO list.

#### 2024-3-7:
In the coming updates, we are working on `deployment`, `fine-tuning`, and `few-shot support`.

**Comment by taofuyu** [comment]
torch.einsum() should be replaced by torch.matmul() and torch.sum(), because einsum() is not supported by most edge devices.
For example, I rewrite the [code](https://github.com/AILab-CVC/YOLO-World/blob/e425669cba9b81bdc0621952262b4d44665c293f/yolo_world/models/dense_heads/yolo_world_head.py#L77):
`x = torch.einsum('bchw,bkc->bkhw', x, w)`
to
`batch, channel, height, width = x.shape`
`_, k, _ = w.shape`
`x = x.permute(0, 2, 3, 1) # bchw->bhwc`
`x = x.reshape(batch, -1, channel) # bhwc->b(hw)c`
`w = w.permute(0, 2, 1) # bkc->bck`
`x = torch.matmul(x, w)`
`x = x.reshape(batch, height, width, k)`
`x = x.permute(0, 3, 1, 2)`
Maybe it is ugly, but it can be deployed.
@wondervictor

**Comment by wondervictor** [comment]
@taofuyu Good idea, Got it!

**Comment by mio410** [comment]
@wondervictor May I ask where should I modify if I want to try using the effect of other text encoders, such as changing the text encoder of CLIP to BEIT-3. Thank you!

**Comment by wondervictor** [comment]
@mio410 Good idea, we do plan to use better and stronger text encoders (e.g., CLIP-Large) now and we are queuing for computation resources to pre-train it. BEIT-3 is a good choice and we are considering it. BTW, what model size are you most in need of currently? I can prioritize that.

**Comment by mio410** [comment]
> @wondervictor May I ask where should I modify if I want to try using the effect of other text encoders, such as changing the text encoder of CLIP to BEIT-3. Thank you!
Besides, I'd like to try using a CLIP model in a different language to see if I can use prompts in that language for open vocabulary detection. Is this possible?

**Comment by mio410** [comment]
> @mio410 Good idea, we do plan to use better and stronger text encoders (e.g., CLIP-Large) now and we are queuing for computation resources to pre-train it. BEIT-3 is a good choice and we are considering it. BTW, what model size are you most in need of currently? I can prioritize that.

I'm looking forward to your work! If possible, I'd like to try open vocabulary detection in other languages. Could you help me with that?

**Comment by taofuyu** [comment]
> @wondervictor May I ask where should I modify if I want to try using the effect of other text encoders, such as changing the text encoder of CLIP to BEIT-3. Thank you!

[here](https://github.com/AILab-CVC/YOLO-World/blob/e425669cba9b81bdc0621952262b4d44665c293f/yolo_world/models/backbones/mm_backbone.py#L205)

**Comment by dikapiliao1** [comment]
Yolo World is based on the word embedding of clip for reparameterization. If we could replace clip with a larger model similar to ChatGPT4, would it understand more？ similar to Sora's powerful ability to understand images.

**Comment by wondervictor** [comment]
> Yolo World is based on the word embedding of clip for reparameterization. If we could replace clip with a larger model similar to ChatGPT4, would it understand more？ similar to Sora's powerful ability to understand images.

Hi @dikapiliao1, it's a nice idea and we plan to do it.

**Comment by xianhonghuang** [comment]
如果我想要更改不同的視覺的backbone要在哪裡可以更改?

**Comment by wondervictor** [comment]
> 如果我想要更改不同的視覺的backbone要在哪裡可以更改?

@xianhonghuang replace the `image_model` config according to your demand:

```python
backbone=dict(
    _delete_=True,
    type='MultiModalYOLOBackbone',
    image_model={{_base_.model.backbone}},
    text_model=dict(
        type='HuggingCLIPLanguageBackbone',
        model_name=text_model_name,
        frozen_modules=['all'])),
```

**Comment by xianhonghuang** [comment]
> > 如果我想更改不同的主幹線要在哪裡可以更改？
> 
> @xianhonghuang`image_model`根據您的需求替換配置：
> 
> ```python
> backbone=dict(
>     _delete_=True,
>     type='MultiModalYOLOBackbone',
>     image_model={{_base_.model.backbone}},
>     text_model=dict(
>         type='HuggingCLIPLanguageBackbone',
>         model_name=text_model_name,
>         frozen_modules=['all'])),
> ```

像是更改_base_ = ('../../third_party/mmyolo/configs/yolov8/'
          'yolov8_l_syncbn_fast_8xb16-500e_coco.py')這部分嗎?

**Comment by wondervictor** [comment]
Hi @xianhonghuang, you can directly override the backbone dictionary configs, e.g., change it to YOLOv7Backbone. BTW, it's suggested to open a new issue to discuss this question and this issue aims for new features and suggestions.

**Comment by RudyCheng** [comment]
config：yolo_world_v2_xl_vlpan_bn_2e-3_100e_4x8gpus_obj365v1_goldg_train_lvis_minival.py is not suit for its model weights

**Comment by wondervictor** [comment]
@RudyCheng, it has been resolved.

**Comment by xiyuan27** [comment]
[target detection on document images], Are there any specialized optimization strategies or support for target detection in vertical domains, specifically for document images such as invoices and passports?

**Comment by thgpddl** [comment]
请问为什么image_demo.py输入的text经过","分割后，还要追加一个空字符串呢，加入text=cat,dog,man，经过代码处理后text=cat,dog,man," "

**Comment by spacewalk01** [comment]
any progress on tensorrt implementation? thanks

**Comment by PrinceP** [comment]
@spacewalk01  https://github.com/PrinceP/tensorrt-cpp-for-onnx?tab=readme-ov-file#yolo-world without dynamic batch support

**Comment by myb1314yxy** [comment]
I would like to ask, how should I use yolo-world to implement unknown classes detection in my own dataset, how should the dataset be divided and made, do I need to pre-define all known classes in yaml file?

**Comment by dreamexs** [comment]
The model weight link of YOLO-World-V2.1 cannot be downloaded.[https://github.com/AILab-CVC/YOLO-World/blob/master/docs/update_20250123.md](url)
